tradingstrategy/binance/candlestick.py

```diff
@@ -0,0 +1,221 @@
+"""Get candlestick price and volume data from Binance.
+"""
+
+import requests
+import datetime
+import pandas as pd
+import numpy as np
+import logging
+
+from tradingstrategy.timebucket import TimeBucket
+from pathlib import Path
+
+
+logger = logging.getLogger(__name__)
+
+
+def get_binance_candlestick_data(
+    symbol: str,
+    time_bucket: TimeBucket,
+    start_at: datetime.datetime,
+    end_at: datetime.datetime,
+    force_redownload=False,
+):
+    """Get clean candlestick price and volume data from Binance. If saved, use saved version, else create saved version.
+
+    .. code-block:: python
+        five_min_data = get_binance_candlestick_data("ETHUSDC", TimeBucket.m5, datetime.datetime(2021, 1, 1), datetime.datetime(2021, 4, 1))
+
+    :param symbol:
+        Trading pair symbol E.g. ETHUSDC
+
+    :param interval:
+        Can be one of `1s, 1m, 3m, 5m, 15m, 30m, 1h, 2h, 4h, 6h, 8h, 12h, 1d, 3d, 1w, 1M`
+
+    :param start_at:
+        Start date of the data
+
+    :param end_at:
+        End date of the data
+
+    :param force_redownload:
+        Force redownload of data from Binance and overwrite cached version
+
+    :return:
+        Pandas dataframe with the OHLCV data for the columns and datetimes as the index
+    """
+
+    # to include the end date, we need to add one day
+    end_at = end_at + datetime.timedelta(days=1)
+
+    if not force_redownload:
+        try:
+            return get_binance_data_parquet(symbol, time_bucket, start_at, end_at)
+        except:
+            pass
+
+    params_str = f"symbol={symbol}&interval={time_bucket.value}"
+
+    if start_at:
+        assert (
+            end_at
+        ), "If you specify a start_at, you must also specify an end_at"
+        assert isinstance(
+            start_at, datetime.datetime
+        ), "start_at must be a datetime.datetime object"
+        assert isinstance(
+            end_at, datetime.datetime
+        ), "end_at must be a datetime.datetime object"
+        start_timestamp = int(start_at.timestamp() * 1000)
+        end_timestamp = int(end_at.timestamp() * 1000)
+
+    # generate timestamps for each iteration
+    dates = [start_at]
+    current_date = start_at
+    while current_date < end_at:
+        if (end_at - current_date) / time_bucket.to_timedelta() > 999:
+            dates.append((current_date + time_bucket.to_timedelta() * 999))
+            current_date += time_bucket.to_timedelta() * 999
+        else:
+            dates.append(end_at)
+            current_date = end_at
+
+    timestamps = [int(date.timestamp() * 1000) for date in dates]
+    open_prices, high_prices, low_prices, close_prices, volume, dates = (
+        [],
+        [],
+        [],
+        [],
+        [],
+        [],
+    )
+
+    for i in range(0, len(timestamps) - 1):
```

Comment:
For user friendliness we can add a `tqdm` progress bar that gives a good time estimator in Jupyter Notebooks if data is being downloaded.

An example here:

https://github.com/tradingstrategy-ai/trading-strategy/blob/f07eb4aaf77bfa2956ab6c97cb12a3ade905cd3b/tradingstrategy/transport/jsonl.py#L141